### ISSUE TITLE
Synchronization error 24.04.2020

### ISSUE THREAD

**Issue by d1vanov** [d1vanov/quentier#314]
Weird synchronization error started to occur on the date specified in the title:
```
EDAM protocol version check failed, QEverCloud exception: Error transferring https://www.evernote.com/edam/user - server replied: OK
```
The problem seems to be caused by misbehaving Evernote servers, see the discussion in [robert7/nixnote#151](https://github.com/robert7/nixnote2/issues/151). 

I tried to see if the workaround mentioned there about revoking authentication works but found out that it's currently broken in Quentier :disappointed: Will fix it asap.

**Comment by kgolding** [comment]
When/will this hit the ppa:d1vanov/quentier-development ?

**Comment by d1vanov** [comment]
I'm working on it so hopefully a bit later today.

**Comment by d1vanov** [comment]
Some updates:
1. I fixed the handling of "Revoke authentication" button in Quentier.
2. It didn't help at all because apparently QtKeychain library which I use to securely store authentication token in the system's keychain fails to delete the token from the keychain although it claims to have done it successfully. I've observed it with Plasma 5 and KWallet, maybe other backends would have better luck.
3. I inserted a hack into QEverCloud to ignore that weird error code and message in particular.
4. The next problem was not working call to `checkProtocolVersion` from Evernote API. It continuously returned HTTP code 429 - "Too many requests". So I've just removed that call for the time being - it's not much useful anyway although Evernote's docs recommend to call it to ensure the client and the server are compatible and can talk to each other.

After these measures were taken, sync started to work for me again, even though I did not manage to actually remove the auth token from the keychain. I'd gladly take any feedback on whether these measures would help for anyone else too.

**Comment by d1vanov** [comment]
I've just pushed the code change to development, so builds should be ready in an hour or two.

**Comment by d1vanov** [comment]
What else I can and need to do here:
1. So far revoking the authentication only attempts to remove the token and shard id from the keychain. I should also remove other non-secret persistent authentication info. It's a proper way to do this and it should help even in the face of QtKeychain's failure to delete the token from the keychain.
2. Make sure that any in-memory caches of auth info are properly invalidated when the persistent auth info is cleaned up, to ensure the effect on the next sync attempt.
3. Try to troubleshoot QtKeychain: try using newer version of it, if it still reproduces, create a bug report, collect some defailed info.

**Comment by d1vanov** [comment]
Pushed some changes to development versions of libquentier and Quentier (builds should follow in a couple of hours), it seems I got authentication revoking working as intended despite the fact that QtKeychain fails to remove auth token for me.

**Comment by robert7** [comment]
Hi, 
I did some analysis in https://github.com/robert7/nixnote2/issues/151#issuecomment-621384522 .. see also my screenshoots with the relevant requests..

It seems that the Evernote responds with http code 429 to the UserStore.getUser call and does not return any user data. Thus just ignoring the 429 code, doesn't seem to help.
Http code 429 means "Too Many Requests" which doesn't give any sense to me.

I also did some investigation why the reauth in nixnote seems to work until the app is restarted. It seems the auth dialog sets some cookies, thich are then added to the UserStore.getUser call, this then works (but only until session end which throws out the cookies).

**Comment by d1vanov** [comment]
I had a somewhat similar problem in Quentier - 429 code - but with another `UserStore` call: `checkProtocolVersion`. I ended up removing it for the time being as it wasn't mission critical.  Calls of `NoteStore` seem to work in Quentier during the synchronization regardless of the last authentication time.

The error code I chose to ignore though is most probably not 429. In `QNetworkReply` the error code is 299 if I temember correctly although I didn't check the http status code attribute. But I remember that only after ignoring that weird error I actually started to receive 429 from `UserStore.checkProtocolVersion`.

I'll check your investigation in more depth a bit later. I don't remember right now whether I call `UserStore.getUser` every sync but probably not, I recall in Evernote's sync doc it was recommended to only update it monthly or so.

**Comment by d1vanov** [comment]
I want to test the cookie hypotheses. It can be done as follows: after OAuth is finished, return the cookie for user store URL as a part of OAuth result, in addition to the auth token. Then this cookie can be stored along with the token and used during the next session. Will try to arrange the test in the next few days.

**Comment by robert7** [comment]
> I don't remember right now whether I call UserStore.getUser every sync but probably not, I recall in Evernote's sync doc it was recommended to only update it monthly or so.

ok I may try to change the code to remove the UserStore.getUser call

UnknownContentError QT error is 299, but the http code was 429

btw I'm not really sure with the cookies, it was just the one thing I noticed was different

**Comment by d1vanov** [comment]
Experiments have shown that UserStore API calls start to work properly again if HTTP requests sent to Evernote service contain a particular cookie which is set by Evernote during OAuth procedure. Evernote team has been [notified](https://discussion.evernote.com/topic/124257-calls-to-userstore-from-evernote-api-stopped-working/) of this effect and they are investigating it. As it's not fixed yet, I've implemented retrieval and persistence of this cookie and its propagation to UserStore API calls during each sync. In order to obtain and save the cookie I added a one time force re-authentication for all Evernote accounts on sync attempt. The changes have been pushed to both development and master branches. Development builds are ready, master ones will take a couple of hours to finish.

Besides that change I've also fixed a couple of unrelated sync problems and created a couple of branches with other fixes which I also intend to merge after some testing. But the main issue with sync here should now be taken care of.
